README.md
# ms-project1
(Web Browser Game)
Arkanoid Clone

Inspiration
--------------------
The reason I picked this game to be my milestone project is because it was one of my favorite games I played on my APPLE IIGS computer growing up. It is essentially a Breakout brick type game, but has power-ups (such as having a the ability to shoot the bricks or the ball being able to break through every brick in its path) and some bricks have to be hit more than once to break them. The goal is to first make a basic Breakout game and if I have time I will add these other features.

Objective
--------------------
How you play is there is a ball that will fall down to your paddle and you have to make sure to block the ball from hitting the ground, otherwise you lose. You have 3 lives. When you successfully block the ball, it will bounce off your paddle and bounce off walls and bricks. When bricks are hit, they break. The goal is to break all the bricks on screen.

Controls
--------------------
Left Arrow: Move Left
Right Arrow: Move right
Spacebar: Restart game

Play
--------------------
https://rearl89.github.io/ms-project1/



CHANGELOG.md
b0b0166 (HEAD -> main, origin/main, origin/HEAD) Separated typescript version to branch
2198018 final change to read me
8a4d626 updated read me
ba3ab7d Changed audio file name to match purpose
db518c6 added into music
d17ff3a cleaned code, added brick break noise
9d1ae31 cleaned up code, and renamed some files for accuracy
d46a501 Added win and lose sounds
07e36db changed game page file from index.html, to game.html
ca140a4 added start page animation
ac96881 added emoji faces to lose and win screens
e103025 changed score to increments of 10
9575602 lives deduction works
6728c77 added start page
a7f0426 added win and lose screens
87f18ec added function to randomize ball start direction
42ef8a2 added score count
4ca9ec0 added collision and adjusted bricks
6f3c038 changed brick height and removed 2 rows
6d9518f Changed score font. Updated README
a5dc2f8 Changed score font. Updated README
7827132 changed bricks again
5f3763a changed brick draw method again
068e75c changing brick making method
95f4fd1 changing brick making method
bf0ef71 added another row of bricks and changed colors
cac0db8 Ball bounces off top of paddle instead of below it
af22107 Reset when ball hits floor. Randomized ball path
479501a Paddle movement added + more bricks
3c5bd78 Added ball bounce, paddle, and some bricks
a403a3b Added ball movement. Changed ball color
a0814ea Added ball
fd28a58 Updated style around canvas play screen w/ title
c4f1522 Linked canvas element and made canvas 2d.
b997e5c Added title, score and canvas.
20dcf45 Updated readme
b898d62 Added HTML, CSS and JS files to repo. HTML boilerplate added.
dac64ea Initial commit
(END)
87f18ec added function to randomize ball start direction
42ef8a2 added score count
4ca9ec0 added collision and adjusted bricks
6f3c038 changed brick height and removed 2 rows
6d9518f Changed score font. Updated README
a5dc2f8 Changed score font. Updated README
7827132 changed bricks again
5f3763a changed brick draw method again
068e75c changing brick making method
95f4fd1 changing brick making method
bf0ef71 added another row of bricks and changed colors
cac0db8 Ball bounces off top of paddle instead of below it
af22107 Reset when ball hits floor. Randomized ball path
479501a Paddle movement added + more bricks
3c5bd78 Added ball bounce, paddle, and some bricks
a403a3b Added ball movement. Changed ball color
a0814ea Added ball
fd28a58 Updated style around canvas play screen w/ title
c4f1522 Linked canvas element and made canvas 2d.
b997e5c Added title, score and canvas.
20dcf45 Updated readme
b898d62 Added HTML, CSS and JS files to repo. HTML boilerplate added.
dac64ea Initial commit
 ESCOD
87f18ec added function to randomize ball start direction
42ef8a2 added score count
4ca9ec0 added collision and adjusted bricks
6f3c038 changed brick height and removed 2 rows
6d9518f Changed score font. Updated README
a5dc2f8 Changed score font. Updated README
7827132 changed bricks again
5f3763a changed brick draw method again
068e75c changing brick making method
95f4fd1 changing brick making method
bf0ef71 added another row of bricks and changed colors
cac0db8 Ball bounces off top of paddle instead of below it
af22107 Reset when ball hits floor. Randomized ball path
479501a Paddle movement added + more bricks
3c5bd78 Added ball bounce, paddle, and some bricks
a403a3b Added ball movement. Changed ball color
a0814ea Added ball
fd28a58 Updated style around canvas play screen w/ title
c4f1522 Linked canvas element and made canvas 2d.
b997e5c Added title, score and canvas.
20dcf45 Updated readme
b898d62 Added HTML, CSS and JS files to repo. HTML boilerplate added.
dac64ea Initial commit
 ESCOC
 start direction

cks
 2 rows
ME
ME




changed colors
nstead of below it
omized ball path
cks
ome bricks
l color

 screen w/ title
anvas 2d.


repo. HTML boilerplate added.

 ESCOD
87f18ec added function to randomize ball start direction
42ef8a2 added score count
4ca9ec0 added collision and adjusted bricks
6f3c038 changed brick height and removed 2 rows
6d9518f Changed score font. Updated README
a5dc2f8 Changed score font. Updated README
7827132 changed bricks again
5f3763a changed brick draw method again
068e75c changing brick making method
95f4fd1 changing brick making method
bf0ef71 added another row of bricks and changed colors
cac0db8 Ball bounces off top of paddle instead of below it
af22107 Reset when ball hits floor. Randomized ball path
479501a Paddle movement added + more bricks
3c5bd78 Added ball bounce, paddle, and some bricks
a403a3b Added ball movement. Changed ball color
a0814ea Added ball
fd28a58 Updated style around canvas play screen w/ title
c4f1522 Linked canvas element and made canvas 2d.
b997e5c Added title, score and canvas.
20dcf45 Updated readme
b898d62 Added HTML, CSS and JS files to repo. HTML boilerplate added.
dac64ea Initial commit
(END)


startScript.js
//create canvas
const cvs = document.getElementById("playArea");
const con = cvs.getContext("2d");

//ball size and position
const ballRadius = 3;
let x = cvs.width / 2;
let y = cvs.height - 20;



//movement and speed of ball
let dx = 1;
let dy = -1;

//paddle size and position
const paddleHeight = 5;
const paddleWidth = 45;
const paddleStartHeight = 8;
let paddleXPosition = (cvs.width - paddleWidth) / 2;

//bricks size and position
const brickHeight = 8;
const brickWidth = 25;
const brickCols = 10;
const brickRows = 5;
const brickPadding = 5;
const brickOffsetTop = 2;
const brickOffsetLeft = 2.5;
const bricks = [];
for(let c = 0; c < brickCols; c++) {
    bricks[c] = [];
    for(let r = 0; r < brickRows; r++) {
        bricks[c][r] = {x: 0, y: 0, status: 1};
    }
}

//collision with bricks
function brickCollision() {
    for(let c = 0; c < brickCols; c++) {
        for(let r = 0; r < brickRows; r++){
            const b = bricks[c][r];
            if (b.status === 1){
                if (x > b.x && x < b.x + brickWidth && y > b.y && y < b.y + brickHeight) {
                    dy = -dy;
                    b.status = 0;
                }
            }
        }
    }
}

//collision with walls and paddle and lose screen
function collision() {
    if(x + dx > cvs.width - ballRadius || x + dx < ballRadius) { //ball bounce off walls
        dx = -dx;
    }else if(y + dy < ballRadius) {
        dy = -dy;
    }else if(y + dy > (cvs.height - ballRadius) - paddleHeight) { //ball bounce off paddle
            dy = -dy;
    }
}

function paddleFollow() {
    paddleXPosition = x - 22;
}

//draw ball
function drawBall() {
    con.beginPath();
    con.arc(x, y, ballRadius, 0, Math.PI * 2);
    con.fillStyle = 'aqua';
    con.fill();
    con.closePath();
}

//draw paddle
function drawPaddle() {
    con.beginPath();
    con.rect(paddleXPosition, cvs.height - paddleStartHeight, paddleWidth, paddleHeight);
    con.fillStyle = '#CFD0D2';
    con.fill();
    con.closePath();
}

//draw bricks
function drawBricks() {
    for(let c = 0; c < brickCols; c++) {
        for(let r = 0; r < brickRows; r++) {
            if(bricks[c][r].status === 1) {
                const brickXPosition = c * (brickWidth + brickPadding) + brickOffsetLeft;
                const brickYPosition = r * (brickHeight + brickPadding) + brickOffsetTop;
                bricks[c][r].x = brickXPosition;
                bricks[c][r].y = brickYPosition;
                con.beginPath();
                con.rect(brickXPosition, brickYPosition, brickWidth, brickHeight);
                con.fillStyle = 'red';
                con.fill();
                con.closePath();
            }
        }
    }
}

//the draw call
function draw() {
    con.clearRect(0, 0, cvs.width, cvs.height); //clears ball trail
    drawBricks();
    drawBall();
    brickCollision();
    collision();
    drawPaddle();
    paddleFollow();
    x += dx;
    y += dy;

}
const interval = setInterval(draw, 10);

src/gameScript.js
//create canvas
var cvs = document.getElementById("playArea");
var con = cvs.getContext("2d");
var livesDisplay = document.querySelector('#lives');
var lives = 3;
var scoreDisplay = document.querySelector('#score');
var score = 0;
//sound
var LOSE = new Audio();
LOSE.src = "sounds/lose.wav";
var WIN = new Audio();
WIN.src = "sounds/win.wav";
var BRICK_BREAK = new Audio();
BRICK_BREAK.src = "sounds/brick_break.wav";
//ball size and position
var ballRadius = 3;
var x = cvs.width / 2;
var y = cvs.height - 20;
//randomly chooses the direction of ball
function randomNum() {
    var num = Math.random();
    if (num < 0.25) {
        return -2;
    }
    else if (num > 0.24 && num < 0.5) {
        return -1;
    }
    else if (num > 0.49 && num < 0.75) {
        return 1;
    }
    else {
        return 2;
    }
}
//movement and speed of ball
var dx = randomNum();
var dy = -2;
function moveBall() {
    x += dx;
    y += dy;
}
//paddle size and position
var paddleHeight = 5;
var paddleWidth = 45;
var paddleStartHeight = 8;
var paddleXPosition = (cvs.width - paddleWidth) / 2;
//bricks size and position
var brickHeight = 8;
var brickWidth = 25;
var brickCols = 10;
var brickRows = 5;
var brickPadding = 5;
var brickOffsetTop = 2;
var brickOffsetLeft = 2.5;
var bricks = [];
for (var c = 0; c < brickCols; c++) {
    bricks[c] = [];
    for (var r = 0; r < brickRows; r++) {
        bricks[c][r] = { x: 0, y: 0, status: 1 };
    }
}
//controls
var rightPressed = false;
var leftPressed = false;
document.addEventListener('keydown', pressKey, false);
document.addEventListener('keyup', releaseKey, false);
function pressKey(e) {
    if (e.key === 'ArrowRight') {
        rightPressed = true;
    }
    else if (e.key === 'ArrowLeft') {
        leftPressed = true;
    }
}
function releaseKey(e) {
    if (e.key === 'ArrowRight') {
        rightPressed = false;
    }
    else if (e.key === 'ArrowLeft') {
        leftPressed = false;
    }
}
//move paddle
function movePaddle() {
    if (rightPressed) {
        paddleXPosition = Math.min(paddleXPosition + 4, cvs.width - paddleWidth);
    }
    else if (leftPressed) {
        paddleXPosition = Math.max(paddleXPosition - 4, 0);
    }
}
//restart game
document.addEventListener('keydown', reloadGame);
function reloadGame(e) {
    if (e.key === " ") {
        document.location.reload();
    }
}
//collision with bricks
function brickCollision() {
    for (var c = 0; c < brickCols; c++) {
        for (var r = 0; r < brickRows; r++) {
            var b = bricks[c][r];
            if (b.status === 1) {
                if (x > b.x && x < b.x + brickWidth && y > b.y && y < b.y + brickHeight) {
                    BRICK_BREAK.play();
                    dy = -dy;
                    score += 10;
                    scoreDisplay.innerHTML = score.toString();
                    b.status = 0;
                    if (score == (brickRows * brickCols) * 10) {
                        endScreen();
                        winText();
                        restartText();
                        WIN.play();
                        clearInterval(interval);
                        reloadGame();
                    }
                }
            }
        }
    }
}
//win and lose displays
function endScreen() {
    con.beginPath();
    con.rect(0, 0, cvs.width, cvs.height);
    con.fillStyle = 'black';
    con.fill();
    con.closePath();
}
function winText() {
    con.font = '16px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText('You win! 😄 Score: ' + score, (cvs.width / 2) - 105, (cvs.height / 2) + 8);
}
function loseText() {
    con.font = '16px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText("You lose! ☹️ Score: " + score, (cvs.width / 2) - 110, (cvs.height / 2) + 8);
}
function restartText() {
    con.font = '8px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText('Press spacebar to restart', (cvs.width / 2) - 65, (cvs.height / 2) + 20);
}
//collision with walls and paddle and lose screen
function collision() {
    if (x + dx > cvs.width - ballRadius || x + dx < ballRadius) { //ball bounce off walls
        dx = -dx;
    }
    else if (y + dy < ballRadius) {
        dy = -dy;
    }
    else if (y + dy > (cvs.height - ballRadius) - paddleHeight) { //ball bounce off paddle
        if (x > paddleXPosition && x < paddleXPosition + paddleWidth) {
            dy = -dy;
        }
        else {
            lives--;
            livesDisplay.innerHTML = lives.toString();
            if (!lives) {
                endScreen();
                loseText();
                restartText();
                LOSE.play();
                clearInterval(interval);
                reloadGame();
            }
            else {
                x = cvs.width / 2;
                y = cvs.height - 20;
                dx = randomNum();
                dy = -2;
                paddleXPosition = (cvs.width - paddleWidth) / 2;
            }
        }
    }
}
//draw ball
function drawBall() {
    con.beginPath();
    con.arc(x, y, ballRadius, 0, Math.PI * 2);
    con.fillStyle = 'aqua';
    con.fill();
    con.closePath();
}
//draw paddle
function drawPaddle() {
    con.beginPath();
    con.rect(paddleXPosition, cvs.height - paddleStartHeight, paddleWidth, paddleHeight);
    con.fillStyle = '#CFD0D2';
    con.fill();
    con.closePath();
}
//draw bricks
function drawBricks() {
    for (var c = 0; c < brickCols; c++) {
        for (var r = 0; r < brickRows; r++) {
            if (bricks[c][r].status === 1) {
                var brickXPosition = c * (brickWidth + brickPadding) + brickOffsetLeft;
                var brickYPosition = r * (brickHeight + brickPadding) + brickOffsetTop;
                bricks[c][r].x = brickXPosition;
                bricks[c][r].y = brickYPosition;
                con.beginPath();
                con.rect(brickXPosition, brickYPosition, brickWidth, brickHeight);
                con.fillStyle = 'red';
                con.fill();
                con.closePath();
            }
        }
    }
}
//the draw call
function draw() {
    con.clearRect(0, 0, cvs.width, cvs.height); //clears ball trail
    drawBricks();
    drawBall();
    moveBall();
    brickCollision();
    collision();
    drawPaddle();
    movePaddle();
}
var interval = setInterval(draw, 10);


src/gameScript.ts
//create canvas
const cvs = document.getElementById("playArea") as HTMLCanvasElement;
const con = cvs.getContext("2d")!;

const livesDisplay = document.querySelector('#lives') as HTMLElement;
let lives = 3;

const scoreDisplay = document.querySelector('#score') as HTMLElement;
let score = 0;

//sound
const LOSE = new Audio();
LOSE.src = "sounds/lose.wav";
const WIN = new Audio();
WIN.src = "sounds/win.wav";
const BRICK_BREAK = new Audio();
BRICK_BREAK.src = "sounds/brick_break.wav"

//ball size and position
const ballRadius = 3;
let x = cvs.width / 2;
let y = cvs.height - 20;

//randomly chooses the direction of ball
function randomNum(): number {
    let num = Math.random();
    if (num < 0.25) {
        return -2;
    } else if (num > 0.24 && num < 0.5) {
        return -1;
    } else if (num > 0.49 && num < 0.75) {
        return 1;
    } else {
        return 2;
    }
}

//movement and speed of ball
let dx = randomNum();
let dy = -2;

function moveBall() {
    x += dx;
    y += dy;
}

//paddle size and position
const paddleHeight = 5;
const paddleWidth = 45;
const paddleStartHeight = 8;
let paddleXPosition = (cvs.width - paddleWidth) / 2;

//bricks size and position
const brickHeight = 8;
const brickWidth = 25;
const brickCols = 10;
const brickRows = 5;
const brickPadding = 5;
const brickOffsetTop = 2;
const brickOffsetLeft = 2.5;
const bricks: { x: number; y: number; status: number; }[][] = [];
for (let c = 0; c < brickCols; c++) {
    bricks[c] = [];
    for (let r = 0; r < brickRows; r++) {
        bricks[c][r] = { x: 0, y: 0, status: 1 };
    }
}

//controls
let rightPressed = false;
let leftPressed = false;
document.addEventListener('keydown', pressKey, false);
document.addEventListener('keyup', releaseKey, false);

function pressKey(e: KeyboardEvent) {
    if (e.key === 'ArrowRight') {
        rightPressed = true;
    } else if (e.key === 'ArrowLeft') {
        leftPressed = true;
    }
}

function releaseKey(e: KeyboardEvent) {
    if (e.key === 'ArrowRight') {
        rightPressed = false;
    } else if (e.key === 'ArrowLeft') {
        leftPressed = false;
    }
}

//move paddle
function movePaddle() {
    if (rightPressed) {
        paddleXPosition = Math.min(paddleXPosition + 4, cvs.width - paddleWidth);
    } else if (leftPressed) {
        paddleXPosition = Math.max(paddleXPosition - 4, 0);
    }
}

//restart game
document.addEventListener('keydown', reloadGame);

function reloadGame(e: KeyboardEvent) {
    if (e.key === " ") {
        document.location.reload();
    }
}

//collision with bricks
function brickCollision() {
    for (let c = 0; c < brickCols; c++) {
        for (let r = 0; r < brickRows; r++) {
            const b

 = bricks[c][r];
            if (b.status === 1) {
                if (x > b.x && x < b.x + brickWidth && y > b.y && y < b.y + brickHeight) {
                    BRICK_BREAK.play();
                    dy = -dy;
                    score += 10;
                    scoreDisplay.innerHTML = score.toString();
                    b.status = 0;
                    if (score == (brickRows * brickCols) * 10) {
                        endScreen();
                        winText();
                        restartText();
                        WIN.play();
                        clearInterval(interval);
                        reloadGame();
                    }
                }
            }
        }
    }
}

//win and lose displays
function endScreen() {
    con.beginPath();
    con.rect(0, 0, cvs.width, cvs.height);
    con.fillStyle = 'black';
    con.fill();
    con.closePath();
}

function winText() {
    con.font = '16px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText('You win! 😄 Score: ' + score, (cvs.width / 2) - 105, (cvs.height / 2) + 8);
}

function loseText() {
    con.font = '16px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText("You lose! ☹️ Score: " + score, (cvs.width / 2) - 110, (cvs.height / 2) + 8);
}

function restartText() {
    con.font = '8px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText('Press spacebar to restart', (cvs.width / 2) - 65, (cvs.height / 2) + 20)
}

//collision with walls and paddle and lose screen
function collision() {
    if (x + dx > cvs.width - ballRadius || x + dx < ballRadius) { //ball bounce off walls
        dx = -dx;
    } else if (y + dy < ballRadius) {
        dy = -dy;
    } else if (y + dy > (cvs.height - ballRadius) - paddleHeight) { //ball bounce off paddle
        if (x > paddleXPosition && x < paddleXPosition + paddleWidth) {
            dy = -dy;
        } else {
            lives--;
            livesDisplay.innerHTML = lives.toString();
            if (!lives) {
                endScreen();
                loseText();
                restartText();
                LOSE.play();
                clearInterval(interval);
                reloadGame();
            } else {
                x = cvs.width / 2;
                y = cvs.height - 20;
                dx = randomNum();
                dy = -2;
                paddleXPosition = (cvs.width - paddleWidth) / 2;
            }
        }
    }
}

//draw ball
function drawBall() {
    con.beginPath();
    con.arc(x, y, ballRadius, 0, Math.PI * 2);
    con.fillStyle = 'aqua';
    con.fill();
    con.closePath();
}

//draw paddle
function drawPaddle() {
    con.beginPath();
    con.rect(paddleXPosition, cvs.height - paddleStartHeight, paddleWidth, paddleHeight);
    con.fillStyle = '#CFD0D2';
    con.fill();
    con.closePath();
}

//draw bricks
function drawBricks() {
    for (let c = 0; c < brickCols; c++) {
        for (let r = 0; r < brickRows; r++) {
            if (bricks[c][r].status === 1) {
                const brickXPosition = c * (brickWidth + brickPadding) + brickOffsetLeft

;
                const brickYPosition = r * (brickHeight + brickPadding) + brickOffsetTop;
                bricks[c][r].x = brickXPosition;
                bricks[c][r].y = brickYPosition;
                con.beginPath();
                con.rect(brickXPosition, brickYPosition, brickWidth, brickHeight);
                con.fillStyle = 'red';
                con.fill();
                con.closePath();
            }
        }
    }
}

//the draw call
function draw() {
    con.clearRect(0, 0, cvs.width, cvs.height); //clears ball trail
    drawBricks();
    drawBall();
    moveBall();
    brickCollision();
    collision();
    drawPaddle();
    movePaddle();
}
const interval = setInterval(draw, 10);

src/indexScript.js
var cvs = document.getElementById("playArea");
var con = cvs.getContext("2d");
var INTRO = new Audio();
INTRO.src = "sounds/intro.wav";
var ballRadius = 3;
var x = cvs.width / 2;
var y = cvs.height - 20;
var dx = 1;
var dy = -1;
function moveBall() {
    x += dx;
    y += dy;
}
var paddleHeight = 5;
var paddleWidth = 45;
var paddleStartHeight = 8;
var paddleXPosition = (cvs.width - paddleWidth) / 2;
var brickHeight = 8;
var brickWidth = 25;
var brickCols = 10;
var brickRows = 5;
var brickPadding = 5;
var brickOffsetTop = 2;
var brickOffsetLeft = 2.5;
var bricks = [];
for (var c = 0; c < brickCols; c++) {
    bricks[c] = [];
    for (var r = 0; r < brickRows; r++) {
        bricks[c][r] = { x: 0, y: 0, status: 1 };
    }
}
function brickCollision() {
    for (var c = 0; c < brickCols; c++) {
        for (var r = 0; r < brickRows; r++) {
            var b = bricks[c][r];
            if (b.status === 1) {
                if (x > b.x && x < b.x + brickWidth && y > b.y && y < b.y + brickHeight) {
                    dy = -dy;
                    b.status = 0;
                }
            }
        }
    }
}
function collision() {
    if (x + dx > cvs.width - ballRadius || x + dx < ballRadius) {
        // Ball bounce off walls
        dx = -dx;
    }
    else if (y + dy < ballRadius) {
        dy = -dy;
    }
    else if (y + dy > cvs.height - ballRadius - paddleHeight) {
        // Ball bounce off paddle
        dy = -dy;
    }
}
// Computer plays
function paddleFollow() {
    paddleXPosition = x - 22;
}
function drawBall() {
    con.beginPath();
    con.arc(x, y, ballRadius, 0, Math.PI * 2);
    con.fillStyle = 'aqua';
    con.fill();
    con.closePath();
}
function drawPaddle() {
    con.beginPath();
    con.rect(paddleXPosition, cvs.height - paddleStartHeight, paddleWidth, paddleHeight);
    con.fillStyle = '#CFD0D2';
    con.fill();
    con.closePath();
}
function drawBricks() {
    for (var c = 0; c < brickCols; c++) {
        for (var r = 0; r < brickRows; r++) {
            if (bricks[c][r].status === 1) {
                var brickXPosition = c * (brickWidth + brickPadding) + brickOffsetLeft;
                var brickYPosition = r * (brickHeight + brickPadding) + brickOffsetTop;
                bricks[c][r].x = brickXPosition;
                bricks[c][r].y = brickYPosition;
                con.beginPath();
                con.rect(brickXPosition, brickYPosition, brickWidth, brickHeight);
                con.fillStyle = 'red';
                con.fill();
                con.closePath();
            }
        }
    }
}
function draw() {
    con.clearRect(0, 0, cvs.width, cvs.height);
    drawBricks();
    drawBall();
    moveBall();
    brickCollision();
    collision();
    drawPaddle();
    paddleFollow();
    INTRO.play();
}
var interval = setInterval(draw, 10);


src/indexScript.ts
const cvs = document.getElementById("playArea") as HTMLCanvasElement;
const con = cvs.getContext("2d") as CanvasRenderingContext2D;

const INTRO = new Audio();
INTRO.src = "sounds/intro.wav";

const ballRadius = 3;
let x = cvs.width / 2;
let y = cvs.height - 20;

let dx = 1;
let dy = -1;

function moveBall() {
  x += dx;
  y += dy;
}

const paddleHeight = 5;
const paddleWidth = 45;
const paddleStartHeight = 8;
let paddleXPosition = (cvs.width - paddleWidth) / 2;

const brickHeight = 8;
const brickWidth = 25;
const brickCols = 10;
const brickRows = 5;
const brickPadding = 5;
const brickOffsetTop = 2;
const brickOffsetLeft = 2.5;
const bricks: { x: number; y: number; status: number }[][] = [];

for (let c = 0; c < brickCols; c++) {
  bricks[c] = [];
  for (let r = 0; r < brickRows; r++) {
    bricks[c][r] = { x: 0, y: 0, status: 1 };
  }
}

function brickCollision() {
  for (let c = 0; c < brickCols; c++) {
    for (let r = 0; r < brickRows; r++) {
      const b = bricks[c][r];
      if (b.status === 1) {
        if (x > b.x && x < b.x + brickWidth && y > b.y && y < b.y + brickHeight) {
          dy = -dy;
          b.status = 0;
        }
      }
    }
  }
}

function collision() {
  if (x + dx > cvs.width - ballRadius || x + dx < ballRadius) {
    // Ball bounce off walls
    dx = -dx;
  } else if (y + dy < ballRadius) {
    dy = -dy;
  } else if (y + dy > cvs.height - ballRadius - paddleHeight) {
    // Ball bounce off paddle
    dy = -dy;
  }
}

// Computer plays
function paddleFollow() {
  paddleXPosition = x - 22;
}

function drawBall() {
  con.beginPath();
  con.arc(x, y, ballRadius, 0, Math.PI * 2);
  con.fillStyle = 'aqua';
  con.fill();
  con.closePath();
}

function drawPaddle() {
  con.beginPath();
  con.rect(paddleXPosition, cvs.height - paddleStartHeight, paddleWidth, paddleHeight);
  con.fillStyle = '#CFD0D2';
  con.fill();
  con.closePath();
}

function drawBricks() {
  for (let c = 0; c < brickCols; c++) {
    for (let r = 0; r < brickRows; r++) {
      if (bricks[c][r].status === 1) {
        const brickXPosition = c * (brickWidth + brickPadding) + brickOffsetLeft;
        const brickYPosition = r * (brickHeight + brickPadding) + brickOffsetTop;
        bricks[c][r].x = brickXPosition;
        bricks[c][r].y = brickYPosition;
        con.beginPath();
        con.rect(brickXPosition, brickYPosition, brickWidth, brickHeight);
        con.fillStyle = 'red';
        con.fill();
        con.closePath();
      }
    }
  }
}

function draw() {
  con.clearRect(0, 0, cvs.width, cvs.height);
  drawBricks();
  drawBall();
  moveBall();
  brickCollision();
  collision();
  drawPaddle();
  paddleFollow();
  INTRO.play();
}

const interval = setInterval(draw, 10);


script.js
//create canvas
const cvs = document.getElementById("playArea");
const con = cvs.getContext("2d");

const livesDisplay = document.querySelector('#lives');
let lives = 3;

const scoreDisplay = document.querySelector('#score');
let score = 0;

//sound
const LOSE = new Audio();
LOSE.src = "sounds/lose.wav";
const WIN = new Audio();
WIN.src = "sounds/win.wav";

//ball size and position
const ballRadius = 3;
let x = cvs.width / 2;
let y = cvs.height - 20;

//randomly chooses the direction of ball
function randomNum() {
    let num = Math.random();
    if(num < 0.25) {
        return -2;
    }else if(num > 0.24 && num < 0.5) {
        return -1;
    }else if(num > 0.49 && num < 0.75) {
        return 1;
    }else{
        return 2;
    }
}

//movement and speed of ball
let dx = randomNum();
let dy = -2;

//paddle size and position
const paddleHeight = 5;
const paddleWidth = 45;
const paddleStartHeight = 8;
let paddleXPosition = (cvs.width - paddleWidth) / 2;

//bricks size and position
const brickHeight = 8;
const brickWidth = 25;
const brickCols = 10;
const brickRows = 5;
const brickPadding = 5;
const brickOffsetTop = 2;
const brickOffsetLeft = 2.5;
const bricks = [];
for(let c = 0; c < brickCols; c++) {
    bricks[c] = [];
    for(let r = 0; r < brickRows; r++) {
        bricks[c][r] = {x: 0, y: 0, status: 1};
    }
}

//controls
let rightPressed = false;
let leftPressed = false;
document.addEventListener('keydown', pressKey, false);
document.addEventListener('keyup', releaseKey, false);

function pressKey(e) {
    if(e.key === 'Right' || e.key === 'ArrowRight') {
        rightPressed = true;
    }else if(e.key === 'Left' || e.key === 'ArrowLeft') {
        leftPressed = true;
    }
}
function releaseKey(e) {
    if(e.key === 'Right' || e.key === 'ArrowRight') {
        rightPressed = false;
    }else if(e.key === 'Left' || e.key === 'ArrowLeft') {
        leftPressed = false;
    }
}

//restart game
document.addEventListener('keydown', pressSpace);

function pressSpace(e) {
    if(e.key == " ") {
        document.location.reload();
    } 
}

//collision with bricks
function brickCollision() {
    for(let c = 0; c < brickCols; c++) {
        for(let r = 0; r < brickRows; r++){
            const b = bricks[c][r];
            if (b.status === 1){
                if (x > b.x && x < b.x + brickWidth && y > b.y && y < b.y + brickHeight) {
                    dy = -dy;
                    score+= 10;
                    scoreDisplay.innerHTML = score;
                    b.status = 0;
                    if(score == (brickRows*brickCols) * 10) {
                        endScreen();
                        winText();
                        restartText();
                        WIN.play();
                        clearInterval(interval);
                        pressSpace(); 
                    }
                }
            }
        }
    }
}

//win and lose displays
function endScreen() {
    con.beginPath();
    con.rect(0, 0, cvs.width, cvs.height);
    con.fillStyle = 'black';
    con.fill();
    con.closePath;
}

function winText() {
    con.font = '16px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText('You win! 😄 Score: '+score, (cvs.width / 2) - 105, (cvs.height / 2) + 8);
}

function loseText() {
    con.font = '16px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText("You lose! ☹️ Score: "+score, (cvs.width / 2) -110, (cvs.height / 2) + 8);
}

function restartText() {
    con.font = '8px Zen Dots';
    con.fillStyle = '#FFF';
    con.fillText('Press spacebar to restart', (cvs.width / 2) - 65, (cvs.height / 2) + 20)
}

//collision with walls and paddle and lose screen
function collision() {
    if(x + dx > cvs.width - ballRadius || x + dx < ballRadius) { //ball bounce off walls
        dx = -dx;
    }else if(y + dy < ballRadius) {
        dy = -dy;
    }else if(y + dy > (cvs.height - ballRadius) - paddleHeight) { //ball bounce off paddle
        if(x > paddleXPosition && x < paddleXPosition + paddleWidth) {
            dy = -dy;
        }else{
            lives--;
            livesDisplay.innerHTML = lives;
            if(!lives) {
            endScreen();
            loseText();
            restartText();
            LOSE.play();
            clearInterval(interval);
            pressSpace();
            }else{
                x = cvs.width / 2;
                y = cvs.height -20;
                dx = randomNum();
                dy = -2;
                paddleXPosition = (cvs.width - paddleWidth) / 2;
            }
        }
    }
}

//draw ball
function drawBall() {
    con.beginPath();
    con.arc(x, y, ballRadius, 0, Math.PI * 2);
    con.fillStyle = 'aqua';
    con.fill();
    con.closePath();
}

//draw paddle
function drawPaddle() {
    con.beginPath();
    con.rect(paddleXPosition, cvs.height - paddleStartHeight, paddleWidth, paddleHeight);
    con.fillStyle = '#CFD0D2';
    con.fill();
    con.closePath();
}

//draw bricks
function drawBricks() {
    for(let c = 0; c < brickCols; c++) {
        for(let r = 0; r < brickRows; r++) {
            if(bricks[c][r].status === 1) {
                const brickXPosition = c * (brickWidth + brickPadding) + brickOffsetLeft;
                const brickYPosition = r * (brickHeight + brickPadding) + brickOffsetTop;
                bricks[c][r].x = brickXPosition;
                bricks[c][r].y = brickYPosition;
                con.beginPath();
                con.rect(brickXPosition, brickYPosition, brickWidth, brickHeight);
                con.fillStyle = 'red';
                con.fill();
                con.closePath();
            }
        }
    }
}

//the draw call
function draw() {
    con.clearRect(0, 0, cvs.width, cvs.height); //clears ball trail
    drawBricks();
    drawBall();
    brickCollision();
    collision();
    drawPaddle();
    if(rightPressed) {
        paddleXPosition = Math.min(paddleXPosition + 4, cvs.width - paddleWidth);
    }else if (leftPressed) {
        paddleXPosition = Math.max(paddleXPosition - 4, 0);
    }
    x += dx;
    y += dy;
}
const interval = setInterval(draw, 10);



